File: 조준장/level2/미로 탈출.py
from collections import deque

def solution(maps):
    
    def bfs(a, b, target):
        dx = [1, -1, 0, 0]
        dy = [0, 0, 1, -1]
        visited = [[-1 for _ in range(m)] for _ in range(n)]
        visited[a][b] = 0
        queue = deque([[a, b]])
        while queue:
            x, y = queue.popleft()
        
            if maps[x][y] == target:
                return visited[x][y]
        
            for i in range(4):
                nx = x + dx[i]
                ny = y + dy[i]
            
                if 0 <= nx < n and 0 <= ny < m and maps[nx][ny] != "X":
                    if visited[nx][ny] == -1:
                        visited[nx][ny] = visited[x][y] + 1
                        queue.append([nx, ny])
        
        return -1

    answer = 0
    n = len(maps)
    m = len(maps[0])
    for i in range(n):
        for j in range(m):
            if maps[i][j] == "S":
                start = [i, j]
            
            elif maps[i][j] == "L":
                lever = [i, j]
    
    start_to_lever = bfs(start[0], start[1], "L")
    if start_to_lever == -1:
        return -1
    
    lever_to_end = bfs(lever[0], lever[1], "E")
    if lever_to_end == -1:
        return -1 
        
    return start_to_lever+lever_to_end

File: 조준장/level2/미로탈출.kt

import java.util.*

class Solution {
    data class Node(val x: Int, val y: Int, val time: Int = 0)

    var n = 0
    var m = 0
    var graph: List<List<String>> = emptyList()
    fun solution(maps: Array<String>): Int {
        var answer: Int = 0
        graph = maps.map { map -> map.map { it.toString() } }
        n = graph.size
        m = graph.first().size
        var s = Node(0, 0)
        var e = Node(0, 0)
        var l = Node(0, 0)

        graph.forEachIndexed { i, items ->
            items.forEachIndexed { j, item ->
                if (item == "S") {
                    s = Node(i, j)
                }

                if (item == "E") {
                    e = Node(i, j)
                }

                if (item == "L") {
                    l = Node(i, j)
                }
            }
        }

        val sl = bfs(s, l)
        if (sl == -1) return -1
        val le = bfs(l, e)
        if (le == -1) return -1

        return sl + le
    }


    val dx = arrayOf(1, -1, 0, 0)
    val dy = arrayOf(0, 0, 1, -1)


    fun bfs(s: Node, e: Node): Int {
        val queue = PriorityQueue<Node>() { o1, o2 ->
            o1.time - o2.time
        }
        queue.add(s)
        
        val visited = Array(n) { Array(m) { false } }
        visited[s.x][s.y] = true

        while (queue.isNotEmpty()) {
            val (cx, cy, ct) = queue.poll()

            if (cx == e.x && cy == e.y) {
                return ct
            }

            for (i in 0..3) {
                val nx = cx + dx[i]
                val ny = cy + dy[i]

                if (nx in 0..n - 1 && ny in 0..m - 1) {
                    if (!visited[nx][ny] && graph[nx][ny] != "X") {
                        visited[nx][ny] = true
                        queue.add(Node(nx, ny, ct + 1))
                    }
                }

            }
        }

        return -1
    }
}
